src/theory/arrays/theory_arrays.cpp

```diff
@@ -1285,36 +1303,92 @@ void TheoryArrays::check(Effort e) {
         d_equalityEngine.assertPredicate(fact, true, fact);
         break;
       case kind::NOT:
+        Debug("pf::array") << "Check: kind::NOT" << std::endl;
+        Debug("pf::array") << "Array theory conflicted? (1): " << ( d_conflict ? "YES" : "NO" ) << std::endl;
+
         if (fact[0].getKind() == kind::SELECT) {
           d_equalityEngine.assertPredicate(fact[0], false, fact);
         } else if (!d_equalityEngine.areDisequal(fact[0][0], fact[0][1], false)) {
           // Assert the dis-equality
           d_equalityEngine.assertEquality(fact[0], false, fact);
 
+          Debug("pf::array") << "Array theory conflicted? (2): " << ( d_conflict ? "YES" : "NO" ) << std::endl;
+
           // Apply ArrDiseq Rule if diseq is between arrays
           if(fact[0][0].getType().isArray() && !d_conflict) {
+            if (d_conflict) { Debug("pf::array") << "Entering the skolemization branch" << std::endl; }
+
             NodeManager* nm = NodeManager::currentNM();
             TypeNode indexType = fact[0][0].getType()[0];
-            TNode k = getSkolem(fact,"array_ext_index", indexType, "an extensional lemma index variable from the theory of arrays", false);
-
-            Node ak = nm->mkNode(kind::SELECT, fact[0][0], k);
-            Node bk = nm->mkNode(kind::SELECT, fact[0][1], k);
-            Node eq = ak.eqNode(bk);
-            Node lemma = fact[0].orNode(eq.notNode());
-            if (options::arraysPropagate() > 0 && d_equalityEngine.hasTerm(ak) && d_equalityEngine.hasTerm(bk)) {
-              // Propagate witness disequality - might produce a conflict
-              d_permRef.push_back(lemma);
-              d_equalityEngine.assertEquality(eq, false, lemma, eq::MERGED_ARRAYS_EXT);
-              ++d_numProp;
+
+            if (!d_proofsEnabled) {
+              Debug("pf::array") << "Check: kind::NOT: array theory making a skolem" << std::endl;
+
+              TNode k = getSkolem(fact,"array_ext_index", indexType, "an extensional lemma index variable from the theory of arrays", false);
+
+              if (options::proof()) {
```

Comment:
Why is this guarded by "if (options::proof()) {" instead of PROOF() ? I'm a bit confused.